bpf/flows.c

```diff
@@ -95,59 +151,137 @@ static inline int fill_ip6hdr(struct ipv6hdr *ip, void *data_end, struct flow *f
+            flow_event->metrics.flags = flags;
+            bpf_ringbuf_submit(flow_event, 0);
+            // Defer the deletion of the entry from the map to usespace since it evicts other CPU metrics
+        } else {
+            if (direction == INGRESS) {
+                bpf_map_update_elem(&xflow_metric_map_ingress, &id, my_flow_counters, BPF_EXIST);
+            } else {
+                bpf_map_update_elem(&xflow_metric_map_egress, &id, my_flow_counters, BPF_EXIST);
+            }
+        }
     } else {
-        flow->direction = direction;
-        flow->bytes = skb->len;
-        bpf_ringbuf_submit(flow, 0);
+        flow_metrics new_flow_counter = {
+            .packets = 1, .bytes=skb->len};
+        new_flow_counter.flow_start_ts = current_time;
+        new_flow_counter.last_pkt_ts = current_time;
+        int ret;
+        if (direction == INGRESS) {
+            ret = bpf_map_update_elem(&xflow_metric_map_ingress, &id, &new_flow_counter,
+                                      BPF_NOEXIST);
+        } else {
+            ret = bpf_map_update_elem(&xflow_metric_map_egress, &id, &new_flow_counter,
+                                      BPF_NOEXIST);
+        }
+        if (ret < 0) {
+            /*
+                When the map is full, we have two choices:
+                    1) Send the new flow entry to userspace via ringbuffer,
+                       until an entry is available.
+                    2) Send an existing flow entry (probably least recently used)
+                       to userspace via ringbuffer, delete that entry, and add in the
+                       new flow to the hash map.
+
+                Ofcourse, 2nd step involves more manipulations and
+                       state maintenance, and no guarantee if it will any performance benefit.
+                Hence, taking the first approach here to send the entry to userspace
+                and do nothing in the eBPF datapath and wait for userspace to create more space in the Map.
+
+            */
+
+            flow_event = bpf_ringbuf_reserve(&flows, sizeof(flow_record), 0);
+            if (!flow_event) {
+                return rc;
+            }
+            export_flow_id(&flow_event->flow_key, id, direction);
+            flow_event->metrics = new_flow_counter;
+            bpf_ringbuf_submit(flow_event, 0);
+        }
     }
-    return TC_ACT_OK;
-}
+    return rc;
 
-SEC("tc/ingress_flow_parse")
-static inline int ingress_flow_parse(struct __sk_buff *skb) {
-    return flow_parse(skb, INGRESS);
 }
-
-SEC("tc/egress_flow_parse")
-static inline int egress_flow_parse(struct __sk_buff *skb) {
-    return flow_parse(skb, EGRESS);
+SEC("tc_ingress")
+int ingress_flow_parse (struct __sk_buff *skb) {
+    return flow_monitor(skb, INGRESS);
 }
 
-char __license[] SEC("license") = "GPL";
+SEC("tc_egress")
+int egress_flow_parse (struct __sk_buff *skb) {
+    return flow_monitor(skb, EGRESS);
+}
+char _license[] SEC("license") = "GPL";
```

Comment:
It was previously `__license` (double underscore). Does it change anything, or this is purely informative / conventional?
As a side-note, I started a discussion on slack about licensing. Not sure if we want to use GPL or another GPL-compatible one.

Reply:
@jotak : SEC("license") is the declaration of license, but the variable "char license[]" can be anything. Generally, I noticed "_license" being used (ref :https://github.com/cilium/tetragon/blob/main/bpf/process/bpf_generic_kprobe.c#L17)